src/HappyConnOpener.cc

```diff
@@ -0,0 +1,406 @@
+    // Starting new connection here will alter the this->nextAttemptTime member
+    // which affects the WaitingConnectors ordering.
+    if (queueSubscribed_) {
+        debugs(17, 8, "already subscribed for starting new connection when ready");
+        return;
+    }
+
+    if (master.path == nullptr || spareConnectionNeeded()) {
+        Comm::ConnectionPointer dest = getCandidatePath();
+        if (dest != nullptr)
+            startConnecting(dest);
+    }
+
+    if (Config.happyEyeballs.connect_limit == 0)
+        return; // feature disabled
+
+    // Else check to start a monitoring process to start secondary connections
+    // if required.
+    if (spare.path == nullptr && existCandidatePath()) {
+        debugs(17, 8, "Schedule a new attempt for later");
+        ScheduleConnectionAttempt(HappyConnOpener::Pointer(this));
+    }
+}
+
+/**
+ * Decide where details need to be gathered to correctly describe a persistent connection.
+ * What is needed:
+ *  -  the address/port details about this link
+ *  -  domain name of server at other end of this link (either peer or requested host)
+ */
+void
+HappyConnOpener::PconnPush(Comm::ConnectionPointer &conn, const char *domain)
+{
+    if (conn->getPeer()) {
+        fwdPconnPool->push(conn, NULL);
+    } else {
+        fwdPconnPool->push(conn, domain);
+    }
+}
+
+Comm::ConnectionPointer
+HappyConnOpener::PconnPop(const Comm::ConnectionPointer &dest, const char *domain, bool retriable)
+{
+    // always call shared pool first because we need to close an idle
+    // connection there if we have to use a standby connection.
+    Comm::ConnectionPointer conn = fwdPconnPool->pop(dest, domain, retriable);
+    if (!Comm::IsConnOpen(conn)) {
+        // either there was no pconn to pop or this is not a retriable xaction
+        if (CachePeer *peer = dest->getPeer()) {
+            if (peer->standby.pool)
+                conn = peer->standby.pool->pop(dest, domain, true);
+        }
+    }
+    return conn; // open, closed, or nil
+}
+
+void
+HappyConnOpener::ConnectionClosed(const Comm::ConnectionPointer &conn)
+{
+    fwdPconnPool->noteUses(fd_table[conn->fd].pconn.uses);
+}
+
+void
+HappyConnOpener::ScheduleConnectorsListCheck()
+{
+    HappyConnOpener::Pointer he = FrontOpener();
+    if (he.valid()) {
+        double startAfter = he->nextAttemptTime > current_dtime ? he->nextAttemptTime - current_dtime : 0.001;
+        eventAdd("HappyConnOpener::CheckForConnectionAttempts", CheckForConnectionAttempts, NULL, startAfter, false);
+    }
+}
+
+void
+HappyConnOpener::ScheduleConnectionAttempt(HappyConnOpener::Pointer happy)
+{
+    happy->queueSubscribed_ = true;
+    if (WaitingConnectors.empty()) {
+        WaitingConnectors.push_back(happy);
+        ScheduleConnectorsListCheck(); // Restart queue run
+        return;
+    }
+    // Exist at least one item in list
+    auto it = --WaitingConnectors.end();
+    for (; it != WaitingConnectors.begin(); --it) {
```

Comment:
Yep this is a problem.
However we are start running the queue from the end, and if I am not wrong it is most possible that the object should placed near the end of list.

My sense is that the order affected by the time a new candidate path found or a previous connection try is finished. I do not believe that two queues can solve the problem. We need an other mechanism.

Reply:
> My sense is that the order affected by the time a new candidate path found or a previous connection try is finished.

How is the wait delay (or wake up time) computed for the new item? There may be several cases, but probably not too many, so I hope you can define each one precisely. I am uncomfortable with the words "sense" and "believe" in this context. We should know these formulas and, depending on those formulas, we may be able to optimize or even remove this potentially expensive search. We may even want to adjust the formulas themselves if it helps.

Reply:
> I am uncomfortable with the words "sense" and "believe" in this context
OK, you are right, I make this comment just before leave the office where my mind was not so clear to understand the related code.
The problem is the following:
 - The HappyConnOpener gets the noteCandidatePath and starts the master connection. 
 - It does not have any alternate candidate path so waits for more before schedules a spare connection
 - A new noteCandidatePath received, but before the spare connection time preconditions achieved so it adds it self to WaitingConnectors.

The reason we have to do linear search is the rate the candidate paths are received from DNS server.

Reply:
What if we queue a spare connection opening immediately, even though we do not have the spare address yet? If that scheduled alarm rings before we have the spare address, then we will set some flag so that we do not have to wait when the spare address finally arrives.

Reconfigurations aside, with such a scheme, all additions related to spare connection opening will use the same gap from _now_ (alarm time = now + constant gap) so they will all go to the end of (their dedicated) queue, right?

Reply:
On reconfigurations I think there is problem, because a now+constant1 maybe is less than now+constant2 or greater, we can not know.

Also what about the case the firsts HappyConOpener objects in waitings queue did not received yet a candidate path? We still need to iterate to find the first HappyConnOpener which has a candidate path.

Reply:
Reconfiguration: I do not think it is important in this discussion. We can, for example, leave a search in place so that reconfiguring to lower gap values works correctly or, upon reconfiguration, recalculate the queue with the new gaps in mind.

> Also what about the case the firsts HappyConOpener objects in waitings queue did not received yet a candidate path? We still need to iterate to find the first HappyConnOpener which has a candidate path.

If the scheduled alarm for HappyConOpener X rings before X has the spare path, then X will set some flag so that X does not have to wait when X finally gets the spare path. The rang alarm for X is immediately deleted from the queue as any other rang alarm. After that deletion, the very first alarm remaining in the queue is the next alarm that should ring (immediately or after a wait). No search there. Or did I misunderstood your question?

Reply:
> What about using two queues, one with those with rang alarm (lets say rangWaiters), and the normal queue?

I do not see the point of having a queue of past/happened events:

1. A time-based event should be processed when it is fired. In this particular case, the event means "the gap restrictions for opening a spare connection are gone". It does _not_ mean "open the spare connection now!". Whoever receives that event should be able to remember that gap restrictions are no longer applicable and then either open a spare connection or continue to wait for other preconditions.
2. Since we do not know the order in which the spare connection paths will become available, it would be costly to find the right event in that additional queue of past events -- it may not be the first event in that queue.

Am I missing something that makes an extra queue attractive?

> The checks (schedule with the eventAdd) should not scheduled if the spare connections limit is reached, and should started again when one or more spare connections closed.

Yes, waiting for the number of connections to go down should not be done via eventAdd. It should be done via at-closing-time checks. The code you are probably talking about should not be even _reached_ when there are too many connections -- we should bail earlier, before starting the checks that lead to eventAdd().

Reply:
> What if we queue a spare connection opening immediately, even though we do not have the spare address yet? If that scheduled alarm rings before we have the spare address, then we will set some flag so that we do not have to wait when the spare address finally arrives.

The management of a such queue is a nightmare. Still there is a problem.
- HappyConnOpener starts a master connection
- HappyConnOpener adds imediatelly itself to the queue to schedule an alarm for spare connection
- The master connection fails, before the alarm ringed.

In this case we must:
  - Start a new master connection
  - Remove the HappyConnOpener from queue (using linear search)
  - Add it again to the tail of the queue.